code_tools/2 type_checkers/index.md

```diff
@@ -0,0 +1,570 @@
+
+</details>
+
+2. Annotate the data structures below:
+
+        
+        foo = ["hello", "world"]
+    
+
+    <textarea name="form[q2.1]" rows="1" required=""></textarea>
+
+        
+        bar = [("Martijn", 1), ("Marleen", 2)]
+    
+
+    <textarea name="form[q2.2]" rows="1" required=""></textarea>
+
+        
+        baz = {1: {2: {3: "hello"}}}
+    
+
+    <textarea name="form[q2.3]" rows="1" required=""></textarea>
+
+<details markdown="1"><summary  markdown="span">Type variables</summary>
+
+Generics work through the use of type variables. In Python these variables are provided by `TypeVar` from the `typing` module. Here is how it works:
+
+
+    from typing import TypeVar
+
+    T = TypeVar('T')  # Can be anything
+    N = TypeVar('N', int, float)  # Must be int or float
+
+
+Type variables can be unconstraint, like `T` above. In this case `T` can be any type at all. Or type variables can be constraint, like `N` above. In which case `N` can only be an `int` or a `float`. Type variables can come in place of actual types to create for instance generic functions:
+
+
+    from typing import Iterable, TypeVar
+
+    T = TypeVar('T')
+
+    def first(items: list[T]) -> T:
+        return items[0]
+
+
+`first` will return the first item in the list, but what type is returned is dependent on the list. For instance, if `first` is called like so:
+
+
+    n = first([1,2,3])
+
+
+Then `n` will be of type `int`. Because a `list[int]` is passed in and `T` will take on the form of `int`. `T` is what is ultimately returned from `first` and that is then why `n` is an `int`.
+
+Type variables can be used outside generic data structures too, for instance:
+
+
+    def longest(a: T, b: T) -> T:
+        return a if len(a) >= len(b) else b
+
+
+This function will work for any type T, and it will return that same type.
+
+</details>
+
+3. Annotate the generic function below with a type variable:
+
+        
+        def repeat(x, n):
+            return [x] * n
+    
+
+    <textarea name="form[q3]" rows="2" required=""></textarea>
+
+## Abstract types
+
+<details markdown="1"><summary  markdown="span">Structural subtyping aka duck typing</summary>
+
+So far we have looked at concrete types, such as integers, strings and lists. These types are expressive, you know exactly what you are working with. But, often these concrete types limit design. Take for instance this function:
+
+
+    def sum(items: list[int]) -> int:
+        total = 0
+        for item in items:
+            total += item
+        return item
+
+
+There is no reason this implementation cannot work with other types of data structures. A tuple of integers or a set of integers should work just fine, but the type hint `list[int]` will only accept a concrete `list`. This is quite unpythonic!
```

Comment:
"unpythonic" is moeilijk te lezen, zeker als het de eerste keer ooit is ;-)
Daarom ben ik gewend soms een koppelteken toe te voegen om de leesbaarheid te verhogen "un-pythonic".